specs/CDPSessionIdSupport.md

```diff
@@ -0,0 +1,328 @@
+                CHECK_FAILURE(args->get_ParameterObjectAsJson(&parameterObjectAsJson));
+                std::wstring eventSource;
+                std::wstring eventDetails = parameterObjectAsJson.get();
+                wil::com_ptr<ICoreWebView2DevToolsProtocolEventReceivedEventArgs2> args2;
+                if (SUCCEEDED(args->QueryInterface(IID_PPV_ARGS(&args2))))
+                {
+                    wil::unique_cotaskmem_string sessionId;
+                    CHECK_FAILURE(args2->get_SessionId(&sessionId));
+                    if (sessionId.get() && *sessionId.get())
+                    {
+                        std::wstring targetId = m_devToolsSessionMap[sessionId.get()];
+                        eventSource = m_devToolsTargetDescriptionMap[targetId];
+                    }
+                }
+                // App code to log these events.
+                LogConsoleLogMessage(eventSource, eventDetails);
+                return S_OK;
+            })
+            .Get(),
+        &m_consoleAPICalledToken));
+    receiver.reset();
+  
+    // Track Target and session info via CDP events.
+    CHECK_FAILURE(
+        m_webView->GetDevToolsProtocolEventReceiver(L"Target.attachedToTarget", &receiver));
+    CHECK_FAILURE(receiver->add_DevToolsProtocolEventReceived(
+        Callback<ICoreWebView2DevToolsProtocolEventReceivedEventHandler>(
+            [this](
+                ICoreWebView2* sender,
+                ICoreWebView2DevToolsProtocolEventReceivedEventArgs* args) -> HRESULT
+            {
+                // A new target is attached, add its info to maps.
+                wil::unique_cotaskmem_string jsonMessage;
+                CHECK_FAILURE(args->get_ParameterObjectAsJson(&jsonMessage));
+                std::wstring sessionId = GetJSONStringField(jsonMessage.get(), L"sessionId");
+                std::wstring targetId = GetJSONStringField(jsonMessage.get(), L"targetId");
+                m_devToolsSessionMap[sessionId] = targetId;
+                std::wstring type = GetJSONStringField(jsonMessage.get(), L"type");
+                std::wstring url = GetJSONStringField(jsonMessage.get(), L"url");
+                m_devToolsTargetDescriptionMap[targetId] = type + L"," + url;
+                wil::com_ptr<ICoreWebView2_10> webview2 = m_webView.try_query<ICoreWebView2_10>();
+                if (webview2)
+                {
+                    // Auto attach to targets further created from this target.
+                    webview2->CallDevToolsProtocolMethodForSession(
+                        sessionId.c_str(), L"Target.setAutoAttach",
+                        LR"({"autoAttach":true,"waitForDebuggerOnStart":false,"flatten":true})",
+                        nullptr);
+                    // Enable Runtime events for the target to receive Runtime.consoleAPICalled from it.
+                    webview2->CallDevToolsProtocolMethodForSession(
+                        sessionId.c_str(), L"Runtime.enable", L"{}", nullptr);
```

Comment:
Suppose I wanted to do something after `CallDevToolsProtocolMethodForSession` completed; i.e., I actually want the completion callback. Can I go into a nested message loop waiting for the completion callback? Or is the completion callback waiting for me to return from the `attachedToTarget` handler to return? (In which case, I need to post this work back to my main thread to run outside of an event handler, and then I have race conditions to worry about.)

Reply:
As with all WebView APIs, interaction with the WebView all happens in the same thread and nested message loop is not supported. Calling APIs from different thread is rejected with UI_E_WRONG_THREAD and event handlers are invoked on the same thread when webview is created. If we want to do something after an event, we would do it in the completed event handler. As all the interaction happens on the same thread, no race condition to worry about.

Reply:
No change required.